src/fcast.c

```diff
@@ -15,57 +22,54 @@ SEXP fcast(SEXP lhs, SEXP val, SEXP nrowArg, SEXP ncolArg, SEXP idxArg, SEXP fil
     SET_VECTOR_ELT(ans, i, VECTOR_ELT(lhs, i));
   }
   // get val cols
+  bool some_fill = any_NA_int(nrows*ncols, idx);
   for (int i=0; i<nval; ++i) {
     SEXP thiscol = VECTOR_ELT(val, i);
     SEXP thisfill = fill;
     SEXPTYPE thistype = TYPEOF(thiscol);
     int nprotect = 0;
-    if (isNull(fill)) {
-      if (LOGICAL(is_agg)[0]) {
-        thisfill = PROTECT(allocNAVector(thistype, 1)); nprotect++;
-      } else thisfill = VECTOR_ELT(fill_d, i);
-    }
-    if (TYPEOF(thisfill) != thistype) {
-      thisfill = PROTECT(coerceVector(thisfill, thistype)); nprotect++;
+    if(some_fill){
+      if (isNull(fill)) {
+	if (LOGICAL(is_agg)[0]) {
+	  thisfill = PROTECT(allocNAVector(thistype, 1)); nprotect++;
+	} else thisfill = VECTOR_ELT(fill_d, i);
+      }
+      if (TYPEOF(thisfill) != thistype) {
+	thisfill = PROTECT(coerceVector(thisfill, thistype)); nprotect++;
+      }
     }
     switch (thistype) {
     case INTSXP:
     case LGLSXP: {
-      const int *ithiscol = INTEGER(thiscol);
-      const int *ithisfill = INTEGER(thisfill);
```

Comment:
These are here like this to avoid the small overhead of the R API call `INTEGER()`. When that call (or the similar ones `REAL()`, `COMPLEX()`, etc) appear inside a loop body or loop condition, there is a noticeable slowdown. R doesn't use macros anymore for those API calls. Even though they are `inline` functions they are slightly heavier inside them due to the welcome ALTREP initiative. Perhaps the compiler is clever enough to branch-predict away the branches inside the inline functions, but that wasn't my experience. A few years ago we swept through the codebase and removed R API calls from loops.

Reply:
This is interesting, I will investigate.

Reply:
I did some time/memory benchmarks to compare usage of INTEGER before for loop, or in each iteration of for loop, source code: https://github.com/tdhock/atime/blob/dt-dcast/vignettes/data.table.Rmd
I observed that there is no significant difference in time/memory usage, see figures below. (median line, min/max ribbon over 10 timings for each input data size N) These data indicate that it does not matter if we use INTEGER before or inside the for loop, and that the speed/memory usage of this branch is comparable to master/release (no significant speedups nor slowdowns).
![pr5549-no-agg-no-missing](https://user-images.githubusercontent.com/932850/205747684-67a20842-073d-4cea-87f5-ee0c6afb0cab.png)
![pr5549-no-agg-half-missing](https://user-images.githubusercontent.com/932850/205747689-3f87e495-529d-4454-9c11-a27b37a1304c.png)
![pr5549-min-agg](https://user-images.githubusercontent.com/932850/205747815-614b3145-3512-4a3a-af09-29b2db702859.png)

Reply:
It would be much easier to understand the code if that would be just simple system.time()

Reply:
Just looking at the last picture it seems that there is noticable difference. It is log y scale there, if you make it non log then it will be very clear. Also it make sense to benchmark only such loops rather than complete dcast call, as we are discussing those macros in general, rather than dcast in particular.

Reply:
Sorry that the code is not just a simple system.time -- this is necessary if we want to see the asymptotic scaling -- if there are any time differences as data size increases.
Actually there is no significant difference here (all median lines are overlapping min/max bands, over the 10 timings). 
If there was a difference on these log log plots it would look something like this https://github.com/Rdatatable/data.table/issues/5424#issuecomment-1197279832

Reply:
Could you try running following code using `dataptr-loop` branch? for me macros are indeed much slower...
```r
test.dataptr = data.table:::dataptrLoop
set.seed(108)
x = 1:1e8
x[2] <- 3L ## ensure no altrep
system.time(test.dataptr(x, 1L)) ## dataptr outside the loop
#   user  system elapsed 
#  0.053   0.000   0.053 
system.time(test.dataptr(x, 2L)) ## dataptr inside the loop
#   user  system elapsed 
#  0.238   0.000   0.239 
```

Reply:
that is interesting thanks @jangorecki I will try that.
it seems to me that these data indicate that INTEGER before the for loop may result in speedups in some cases, but in the context of dcast it does not make a difference (other parts of the dcast code are the bottleneck, not this particular loop with INTEGER)

Reply:
![pr5549-INTEGER-loop-alone](https://user-images.githubusercontent.com/932850/206268138-a0dddcdc-3e80-470e-a515-7e6b20b50198.png)
figure above is benchmark for Jan's test.dataptr, figure below is for two versions of dcast.
![pr5549-fcast_in](https://user-images.githubusercontent.com/932850/206268142-9fbe4d8a-5b1f-497b-afe9-05136f00eaf2.png)
These data indicate that there is some speedup if you move the INTEGER call before the loop, but using that technique does not result in speedups for dcast, because this loop is not the bottleneck.

Reply:
About Michael's point, I tried these benchmarks on both a windows and a linux machine, and got very similar results (windows results reported here).